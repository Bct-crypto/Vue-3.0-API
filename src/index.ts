import { useReducer, useEffect, Dispatch } from 'react'
import { render } from 'react-dom'
import objectPath from 'object-path'
import immutable from 'object-path-immutable'

interface IStringMap<T> {
  [key: string]: T,
}

type Action = {
  type: 'set' | 'del',
  path: Array<string>,
  value?: any,
}

function reducer(state: IStringMap<any>, action: Action) {
  switch (action.type) {
    case 'set':
      return immutable.set(state, action.path, action.value)
    case 'del':
      return immutable.del(state, action.path)
    default:
      throw new Error(`reducer not support ${action.type} now!`)
  }
}

function useState<T extends IStringMap<any>>(obj: T): {[key in keyof T]: any} {
  const [state, dispatch] = useReducer(reducer, obj);

  return doUseState(state, dispatch, obj, [])
}

function doUseState<T extends IStringMap<any>, U extends IStringMap<any>>(state: U, dispatch: Dispatch<Action>, obj: T, path: Array<string>): {[key in keyof T]: any} {
  const descriptors: PropertyDescriptorMap = Object.keys(obj).reduce((p, c) => {
    const origV = obj[c]

    if (Object.prototype.toString.call(origV) === '[object Object]') {
      const v = doUseState(state, dispatch, origV, [...path, c])
      return {
        ...p,
        [c]: {
          get() {
            return v
          }
        }
      }
    }

    return {
      ...p,
      [c]: {
        get() {
          return objectPath.get(state, [...path, c])
        },
        set(newV: any) {
          dispatch({
            type: 'set',
            path: [...path, c],
            value: newV,
          })
        },
      },
    }
  }, {})

  return Object.defineProperties({}, descriptors)
}

export {
  useState,
  useEffect,
  render,
}
