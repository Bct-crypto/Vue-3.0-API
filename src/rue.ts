import { useState as _useState, useEffect } from 'react'
import { render } from 'react-dom'

interface IStringMap<T> {
  [key: string]: T,
}

function useState<T extends IStringMap<any>>(obj: T): {[key in keyof T]: any} {
  const descriptors: PropertyDescriptorMap = Object.keys(obj).reduce((p, c) => {
    const [v, setV] = _useState(obj[c])
    return {
      ...p,
      [c]: {
        get() {
          return v
        },
        set(newV: any) {
          setV(newV)
        },
      },
    }
  }, {})

  return Object.defineProperties({}, descriptors)
}

export {
  useState,
  useEffect,
  render,
}
