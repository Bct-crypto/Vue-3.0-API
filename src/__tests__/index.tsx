/* eslint-disable react/no-unescaped-entities */
import React from 'react'
import { render, fireEvent, getByTestId } from '@testing-library/react'
import { useState } from '../../src'

const App: React.FC = () => {
    const state = useState({
        count: 0,
        user: {
            name: '',
            age: 0,
        },
        users: [
            {
                name: 'a',
                age: 1,
            },
            {
                name: 'b',
                age: 2,
            },
        ],
    })

    return (
        <div>
            <div>
                count: <span data-testid='count'>{state.count}</span>
            </div>
            <div>
                <span data-testid='username'>{state.user.name}</span>
                <span data-testid='age'>{state.user.age}</span>
            </div>
            <ul>
                {state.users.map((u, idx) => (
                    <li date-testid={`user-${idx}`} key={`user-${idx}`}>
                        <span date-testid={`user-${idx}-name`}>{u.name}</span>
                        <span date-testid={`user-${idx}-age`}>{u.age}</span>
                    </li>
                ))}
            </ul>
            <div>
                <button onClick={() => state.count++}>increment</button>
                <button onClick={() => state.count--}>decrement</button>
            </div>
            <div>
                <input
                    type='text'
                    data-testid='username-ipt'
                    value={state.user.name}
                    onChange={e => (state.user.name = e.target.value)}
                />
                <input
                    type='number'
                    data-testid='age-ipt'
                    value={state.user.age}
                    onChange={e => (state.user.age = parseInt(e.target.value))}
                />
                <button data-testid='add-user-btn' onClick={() => (state.users = [...state.users, { ...state.user }])}>
                    Add
                </button>
            </div>
            <div>
                {state.users.map((u, i) => (
                    <button data-testid={`user-btn-${i}`} key={`user-btn-${i}`} onClick={() => u.age++}>
                        Increase {u.name}'s age
                    </button>
                ))}
            </div>
        </div>
    )
}

test('renders correctly', () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
})

test('increment and decrement', async () => {
    const { asFragment, getByText } = render(<App />)
    const incrBtn = getByText('increment')
    const decrBtn = getByText('decrement')
    fireEvent.click(incrBtn)
    fireEvent.click(incrBtn)
    fireEvent.click(incrBtn)
    expect(asFragment()).toMatchSnapshot()
    fireEvent.click(decrBtn)
    fireEvent.click(decrBtn)
    fireEvent.click(decrBtn)
    expect(asFragment()).toMatchSnapshot()
})

test('change user', async () => {
    const { asFragment, getByTestId } = render(<App />)
    const userNameIpt = getByTestId('username-ipt')
    const ageIpt = getByTestId('age-ipt')
    fireEvent.change(userNameIpt, {
        target: {
            value: 'yetone',
        },
    })
    fireEvent.change(ageIpt, {
        target: {
            value: '21',
        },
    })
    expect(asFragment()).toMatchSnapshot()
})

test('add users', async () => {
    const { asFragment, getByTestId } = render(<App />)
    const userNameIpt = getByTestId('username-ipt')
    const ageIpt = getByTestId('age-ipt')
    const addUserBtn = getByTestId('add-user-btn')
    fireEvent.change(userNameIpt, {
        target: {
            value: 'yetone',
        },
    })
    fireEvent.change(ageIpt, {
        target: {
            value: '21',
        },
    })
    fireEvent.click(addUserBtn)
    fireEvent.change(userNameIpt, {
        target: {
            value: 'yetone1',
        },
    })
    fireEvent.change(ageIpt, {
        target: {
            value: '23',
        },
    })
    fireEvent.click(addUserBtn)
    expect(asFragment()).toMatchSnapshot()
    const userBtn0 = getByTestId('user-btn-0')
    const userBtn1 = getByTestId('user-btn-1')
    const userBtn2 = getByTestId('user-btn-2')
    const userBtn3 = getByTestId('user-btn-3')
    fireEvent.click(userBtn0)
    fireEvent.click(userBtn1)
    fireEvent.click(userBtn1)
    fireEvent.click(userBtn2)
    fireEvent.click(userBtn2)
    fireEvent.click(userBtn2)
    fireEvent.click(userBtn3)
    fireEvent.click(userBtn3)
    fireEvent.click(userBtn3)
    fireEvent.click(userBtn3)
    expect(asFragment()).toMatchSnapshot()
})
