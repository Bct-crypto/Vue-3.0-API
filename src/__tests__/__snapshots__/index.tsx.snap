// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add users 1`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        0
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      >
        yetone1
      </span>
      <span
        data-testid="age"
      >
        23
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          1
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          2
        </span>
      </li>
      <li
        date-testid="user-2"
      >
        <span
          date-testid="user-2-name"
        >
          yetone
        </span>
        <span
          date-testid="user-2-age"
        >
          21
        </span>
      </li>
      <li
        date-testid="user-3"
      >
        <span
          date-testid="user-3-name"
        >
          yetone1
        </span>
        <span
          date-testid="user-3-age"
        >
          23
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value="yetone1"
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="23"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
      <button
        data-testid="user-btn-2"
      >
        Increase yetone's age
      </button>
      <button
        data-testid="user-btn-3"
      >
        Increase yetone1's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`add users 2`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        0
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      >
        yetone1
      </span>
      <span
        data-testid="age"
      >
        23
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          2
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          4
        </span>
      </li>
      <li
        date-testid="user-2"
      >
        <span
          date-testid="user-2-name"
        >
          yetone
        </span>
        <span
          date-testid="user-2-age"
        >
          24
        </span>
      </li>
      <li
        date-testid="user-3"
      >
        <span
          date-testid="user-3-name"
        >
          yetone1
        </span>
        <span
          date-testid="user-3-age"
        >
          27
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value="yetone1"
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="23"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
      <button
        data-testid="user-btn-2"
      >
        Increase yetone's age
      </button>
      <button
        data-testid="user-btn-3"
      >
        Increase yetone1's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`change user 1`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        0
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      >
        yetone
      </span>
      <span
        data-testid="age"
      >
        21
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          1
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          2
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value="yetone"
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="21"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`increment and decrement 1`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        3
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      />
      <span
        data-testid="age"
      >
        0
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          1
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          2
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value=""
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="0"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`increment and decrement 2`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        0
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      />
      <span
        data-testid="age"
      >
        0
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          1
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          2
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value=""
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="0"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      count: 
      <span
        data-testid="count"
      >
        0
      </span>
    </div>
    <div>
      <span
        data-testid="username"
      />
      <span
        data-testid="age"
      >
        0
      </span>
    </div>
    <ul>
      <li
        date-testid="user-0"
      >
        <span
          date-testid="user-0-name"
        >
          a
        </span>
        <span
          date-testid="user-0-age"
        >
          1
        </span>
      </li>
      <li
        date-testid="user-1"
      >
        <span
          date-testid="user-1-name"
        >
          b
        </span>
        <span
          date-testid="user-1-age"
        >
          2
        </span>
      </li>
    </ul>
    <div>
      <button>
        increment
      </button>
      <button>
        decrement
      </button>
    </div>
    <div>
      <input
        data-testid="username-ipt"
        type="text"
        value=""
      />
      <input
        data-testid="age-ipt"
        type="number"
        value="0"
      />
      <button
        data-testid="add-user-btn"
      >
        Add
      </button>
    </div>
    <div>
      <button
        data-testid="user-btn-0"
      >
        Increase a's age
      </button>
      <button
        data-testid="user-btn-1"
      >
        Increase b's age
      </button>
    </div>
  </div>
</DocumentFragment>
`;
